Rank shows different results between code-runs

#### What language are you using?
Python

#### Have you tried latest version of polars?

- [yes]

If the problem was resolved, please update polars. :)

#### What version of polars are you using?

master as of yesterday

#### What operating system are you using polars on?
Linux

#### What language version are you using
Python 3.10

#### Describe your bug.
I have found a bug that makes code not reproducible between runs, when I think it should be.

I want to add a numbering to my data, grouped by `"customer_ID"` and numbered according to `"date"`.
The desirable result in the minimal example below is:
```python
shape: (3, 4)
┌─────────────┬──────┬───────┬──────┐
│ customer_ID ┆ date ┆ count ┆ rank │
│ ---         ┆ ---  ┆ ---   ┆ ---  │
│ str         ┆ i64  ┆ i64   ┆ i64  │
╞═════════════╪══════╪═══════╪══════╡
│ 0           ┆ 1    ┆ 2     ┆ 1    │
├╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌┤
│ 0           ┆ 2    ┆ 2     ┆ 2    │
├╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌┤
│ 1           ┆ 3    ┆ 1     ┆ 1    │
└─────────────┴──────┴───────┴──────┘
```

In the following code example, I get a different result only sometimes when calling the exact same code. In the minimal example that I have created, the "wrong" result happens about half the time, while on bigger data, it happened once every ~15 times. I thought I was going crazy. Real schrödinger's bug - it was only there when I wasn't looking for it.

I have narrowed it down to a few things:
- it is probably related to UTF-8 strings (Using ints does not show the error)
- Might be related to using `.over` twice in the same `with_columns`

The minimal example runs by itself, but I've included a slightly larger example (100 rows) if it helps debugging:
[bug.parquet.zip](https://github.com/pola-rs/polars/files/9225872/bug.parquet.zip)

```python
import polars as pl

df = pl.DataFrame(
    {
        "customer_ID": ["0", "0", "1",], # Using ints instead of strings removes the bug. This was originally a looong id string, changed for readability here.
        "date": [1, 2, 3], # this was originally a column of dates, but I found that changing to ints also has the same behaviour
    }
)
# df = pl.read_parquet("bug.parquet") # uncomment this line to load slightly larger data

data = []
for _ in range(50):
    df2 = df.sort(by=["customer_ID", "date"]).with_columns(
        [
            pl.count("date")
            .over(pl.col("customer_ID"))
            .alias("count"),  # Removing the count-chain removes the bug
            pl.col("date")
            .rank(method="ordinal")
            .over(pl.col("customer_ID"))
            .alias("rank"),  # <-- bug appears in this column
        ]
    )
    # If loading from "bug.parquet", uncomment the following line to make finding the line with the bug easier
    # df2 = df2.filter(pl.col("customer_ID") == "00007889e4fcd2614b6cbe7f8f3d2e5c728eca32d9eb8ad51ca8b8c4a24cefed")
    data.append(df2)

for i, (first, second) in enumerate(zip(data, data[1:])):
    assert (
        (first == second).to_numpy().all()
    ), f"Index {i} != {i+1}\n{first}\n\n{second}"



### Error below:
AssertionError: Index 3 != 4
shape: (3, 4)
┌─────────────┬──────┬───────┬──────┐
│ customer_ID ┆ date ┆ count ┆ rank │
│ ---         ┆ ---  ┆ ---   ┆ ---  │
│ str         ┆ i64  ┆ u32   ┆ u32  │
╞═════════════╪══════╪═══════╪══════╡
│ 0           ┆ 1    ┆ 2     ┆ 1    │
├╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌┤
│ 0           ┆ 2    ┆ 2     ┆ 2    │
├╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌┤
│ 1           ┆ 3    ┆ 1     ┆ 1    │
└─────────────┴──────┴───────┴──────┘

shape: (3, 4)
┌─────────────┬──────┬───────┬──────┐
│ customer_ID ┆ date ┆ count ┆ rank │
│ ---         ┆ ---  ┆ ---   ┆ ---  │
│ str         ┆ i64  ┆ u32   ┆ u32  │
╞═════════════╪══════╪═══════╪══════╡
│ 0           ┆ 1    ┆ 2     ┆ 1    │
├╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌┤
│ 0           ┆ 2    ┆ 2     ┆ 1    │
├╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌┤
│ 1           ┆ 3    ┆ 1     ┆ 2    │
└─────────────┴──────┴───────┴──────┘
```

--- a/polars/polars-core/src/frame/groupby/proxy.rs
+++ b/polars/polars-core/src/frame/groupby/proxy.rs
@@ -230,7 +230,11 @@ impl GroupsProxy {
     #[cfg(feature = "private")]
     pub fn sort(&mut self) {
         match self {
-            GroupsProxy::Idx(groups) => groups.sort(),
+            GroupsProxy::Idx(groups) => {
+                if !groups.is_sorted() {
+                    groups.sort()
+                }
+            }
             GroupsProxy::Slice { groups, rolling } => {
                 if !*rolling {
                     groups.sort_unstable_by_key(|[first, _]| *first);

--- a/polars/polars-lazy/src/physical_plan/expressions/window.rs
+++ b/polars/polars-lazy/src/physical_plan/expressions/window.rs
@@ -228,9 +228,8 @@ impl PhysicalExpr for WindowExpr {
             .all(|s| matches!(s.is_sorted(), IsSorted::Ascending | IsSorted::Descending));
         let explicit_list_agg = self.is_explicit_list_agg();
 
-        let create_groups = || {
-            // if we flatten this column we need to make sure the groups are sorted.
-            let sorted = self.options.explode ||
+        // if we flatten this column we need to make sure the groups are sorted.
+        let sort_groups = self.options.explode ||
             // if not
             //      `col().over()`
             // and not
@@ -240,7 +239,9 @@ impl PhysicalExpr for WindowExpr {
             // and keys are sorted
             //  we may optimize with explode call
             (!self.is_simple_column_expr() && !explicit_list_agg && sorted_keys && !self.is_aggregation());
-            let gb = df.groupby_with_series(groupby_columns.clone(), true, sorted)?;
+
+        let create_groups = || {
+            let gb = df.groupby_with_series(groupby_columns.clone(), true, sort_groups)?;
             let out: Result<GroupsProxy> = Ok(gb.take_groups());
             out
         };
@@ -261,6 +262,9 @@ impl PhysicalExpr for WindowExpr {
                 if df.height() > 0 {
                     assert!(!gt.is_empty());
                 };
+                if sort_groups {
+                    gt.sort()
+                }
 
                 // We take now, but it is important that we set this before we return!
                 // a next windows function may get this cached key and get an empty if this

--- a/py-polars/tests/test_window.py
+++ b/py-polars/tests/test_window.py
@@ -169,3 +169,25 @@ def test_count_window() -> None:
         .with_column(pl.count().over("a"))["count"]
         .to_list()
     ) == [2, 2, 1]
+
+
+def test_window_cached_keys_sorted_update_4183() -> None:
+    df = pl.DataFrame(
+        {
+            "customer_ID": [
+                "0",
+                "0",
+                "1",
+            ],
+            "date": [1, 2, 3],
+        }
+    )
+    assert df.sort(by=["customer_ID", "date"]).select(
+        [
+            pl.count("date").over(pl.col("customer_ID")).alias("count"),
+            pl.col("date")
+            .rank(method="ordinal")
+            .over(pl.col("customer_ID"))
+            .alias("rank"),
+        ]
+    ).to_dict(False) == {"count": [2, 2, 1], "rank": [1, 2, 1]}

